### DISCUSSION TITLE
How to design a multi-threaded QUIC server with a single UDP port.

### DISCUSSION THREAD

**Discussion by terryykl** [ngtcp2/ngtcp2#1197] [Q&A]
Hello, I have no more tutorial document for reference apart from the example QUIC server program "qtlsserver". The "Server" object is working on a limited number of UDP sockets as an I/O hub for all existing QUIC connections (each server side QUIC connection is represented by a "Handler" object). To support a huge number of concurrent QUIC connections, we usually need to hash these connections to different threads. But suppose if we only have a single UDP socket (because usually QUIC server has only a single well-known UDP port 443), how can those connections that hashed to a number of different threads work on the single UDP socket? For TCP server, it's no problem because each new incoming TCP connection is identified by a four-tuple, and a distinct TCP socket is created for it. This TCP socket can be hash to a particular thread. I wonder if I can do it some way for QUIC server?
I thought about isolating UDP socket from QUIC connections, which means UDP socket can only be seen by "Server". When the Server receives a UDP datagram, it checks which Handler (QUIC connection) should handle it and send it to that thread. But it entails extra thread switch in UDP datagram handling which is less preferred.
Do you have some suggestion for this case?

**Comment by tatsuhiro-t** [comment]
There are several options.  The architecture I use for nghttpx proxy server in nghttp2 is each thread has listening 443 QUIC port with SO_REUSEPORT.
Each thread has a unique identifier encoded in Destination Connection ID.  ebpf program sk_reuseport inspects the Connection ID, and steer the packet to the correct socket.

**Reply by terryykl** [reply]
Thanks for your prompt reply. My immediate doubt is how to correctly hash the Client's Initial Packet which takes a DCID generated by client. But I think nghttpx code will explain it with more details. I'll be looking at nghttpx's design.
By the way, you mentioned some other options, I'm curious about them.
